File: Sources/DebridStreamer/Services/AI/OllamaProvider.swift
import Foundation

struct OllamaProvider: AIAssistantProvider {
    let kind: AIProviderKind = .ollama

    private let endpoint: URL
    private let model: String
    private let session: URLSession

    init(endpoint: URL, model: String = "llama3.1:8b", session: URLSession = .shared) {
        self.endpoint = endpoint
        self.model = model
        self.session = session
    }

    func recommend(prompt: String, candidateTitles: [String], maxResults: Int) async throws -> AIProviderRecommendationResult {
        let envelope = AIAssistantJSONParser.promptEnvelope(
            userPrompt: prompt,
            candidateTitles: candidateTitles,
            maxResults: maxResults
        )

        let payload = OllamaRequest(
            model: model,
            messages: [.init(role: "user", content: envelope)],
            stream: false
        )

        var request = URLRequest(url: endpoint)
        request.httpMethod = "POST"
        request.timeoutInterval = 60
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(payload)

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw AIAssistantProviderError.invalidResponse
        }
        guard (200...299).contains(http.statusCode) else {
            let errorText = String(data: data, encoding: .utf8) ?? "Ollama error"
            throw AIAssistantProviderError.apiError(errorText)
        }

        let decoded = try JSONDecoder().decode(OllamaResponse.self, from: data)
        guard let content = decoded.message?.content else {
            throw AIAssistantProviderError.invalidResponse
        }
        let recommendations = AIAssistantJSONParser.parseRecommendations(from: content, maxResults: maxResults)
        let inputTokens = AIAssistantJSONParser.estimatedTokenCount(for: envelope)
        let outputTokens = AIAssistantJSONParser.estimatedTokenCount(for: content)
        return AIProviderRecommendationResult(
            model: model,
            recommendations: recommendations,
            rawText: content,
            usage: AIUsageMetrics(
                inputTokens: inputTokens,
                outputTokens: outputTokens,
                totalTokens: inputTokens + outputTokens,
                estimatedCostUSD: 0
            )
        )
    }

    /// Single-shot completion for the NL→filter mood discovery: sends `prompt`
    /// verbatim and returns the raw text (no recommendation envelope).
    func complete(prompt: String) async throws -> String {
        let payload = OllamaRequest(
            model: model,
            messages: [.init(role: "user", content: prompt)],
            stream: false
        )

        var request = URLRequest(url: endpoint)
        request.httpMethod = "POST"
        request.timeoutInterval = 60
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONEncoder().encode(payload)

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw AIAssistantProviderError.invalidResponse
        }
        guard (200...299).contains(http.statusCode) else {
            throw AIAssistantProviderError.apiError(String(data: data, encoding: .utf8) ?? "Ollama error")
        }
        let decoded = try JSONDecoder().decode(OllamaResponse.self, from: data)
        guard let content = decoded.message?.content else {
            throw AIAssistantProviderError.invalidResponse
        }
        return content
    }
}

private struct OllamaRequest: Encodable {
    struct Message: Encodable {
        let role: String
        let content: String
    }

    let model: String
    let messages: [Message]
    let stream: Bool
}

private struct OllamaResponse: Decodable {
    struct Message: Decodable {
        let role: String?
        let content: String
    }

    let message: Message?
}

File: web/src/services/ai/OllamaProvider.ts
// Port of Sources/DebridStreamer/Services/AI/OllamaProvider.swift.
//
// A fetch-based Ollama chat provider. Mirrors the Swift struct: POSTs to the
// caller-supplied endpoint with a non-streaming chat request, reads
// `message.content` as the response text, and - because Ollama returns no token
// usage - fills usage from the local heuristic token counter (estimatedCostUSD
// is always 0 for a local model). No API key. The `fetch` implementation is
// injectable so tests stub the network.

import type {
  AIProviderKind,
  AIProviderRecommendationResult,
} from "./models";
import {
  type AIAnalyzeTitleInput,
  AIAssistantJSONParser,
  AIAssistantProviderError,
  type AIProviderAnalysisResult,
  type AIAssistantProvider,
  type FetchImpl,
  parsePersonalizedAnalysis,
  personalizedAnalysisPrompt,
  resolveFetch,
} from "./types";

// MARK: - Request shape (mirrors OllamaRequest)

interface OllamaRequestBody {
  model: string;
  messages: { role: string; content: string }[];
  stream: boolean;
}

// MARK: - Raw response (mirrors OllamaResponse)

interface RawOllamaResponse {
  message?: { role?: string | null; content: string } | null;
}

export class OllamaProvider implements AIAssistantProvider {
  readonly kind: AIProviderKind = "ollama";

  private readonly endpoint: string;
  private readonly model: string;
  private readonly fetchImpl: FetchImpl;

  constructor(endpoint: string, model = "llama3.1:8b", fetchImpl?: FetchImpl) {
    this.endpoint = endpoint;
    this.model = model;
    this.fetchImpl = resolveFetch(fetchImpl);
  }

  async recommend(
    prompt: string,
    candidateTitles: string[],
    maxResults: number,
  ): Promise<AIProviderRecommendationResult> {
    const envelope = AIAssistantJSONParser.promptEnvelope(
      prompt,
      candidateTitles,
      maxResults,
    );

    const payload: OllamaRequestBody = {
      model: this.model,
      messages: [{ role: "user", content: envelope }],
      stream: false,
    };

    const response = await this.fetchImpl(this.endpoint, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    });

    if (!(response.status >= 200 && response.status <= 299)) {
      const errorText = (await response.text().catch(() => "")) || "Ollama error";
      throw AIAssistantProviderError.apiError(errorText);
    }

    const decoded = JSON.parse(await response.text()) as RawOllamaResponse;
    const content = decoded.message?.content;
    if (content == null) {
      throw AIAssistantProviderError.invalidResponse();
    }

    const recommendations = AIAssistantJSONParser.parseRecommendations(
      content,
      maxResults,
    );
    const inputTokens = AIAssistantJSONParser.estimatedTokenCount(envelope);
    const outputTokens = AIAssistantJSONParser.estimatedTokenCount(content);

    return {
      model: this.model,
      recommendations,
      rawText: content,
      usage: {
        inputTokens,
        outputTokens,
        totalTokens: inputTokens + outputTokens,
        estimatedCostUSD: 0,
      },
    };
  }

  async analyzeTitle(
    input: AIAnalyzeTitleInput,
  ): Promise<AIProviderAnalysisResult> {
    const prompt = personalizedAnalysisPrompt(input);

    const payload: OllamaRequestBody = {
      model: this.model,
      messages: [{ role: "user", content: prompt }],
      stream: false,
    };

    const response = await this.fetchImpl(this.endpoint, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    });

    if (!(response.status >= 200 && response.status <= 299)) {
      const errorText = (await response.text().catch(() => "")) || "Ollama error";
      throw AIAssistantProviderError.apiError(errorText);
    }

    const decoded = JSON.parse(await response.text()) as RawOllamaResponse;
    const content = decoded.message?.content;
    if (content == null) {
      throw AIAssistantProviderError.invalidResponse();
    }

    const analysis = parsePersonalizedAnalysis(content);
    // Ollama returns no token usage - fill from the local heuristic counter
    // (estimatedCostUSD is always 0 for a local model). Mirrors recommend().
    const inputTokens = AIAssistantJSONParser.estimatedTokenCount(prompt);
    const outputTokens = AIAssistantJSONParser.estimatedTokenCount(content);

    return {
      model: this.model,
      analysis,
      rawText: content,
      usage: {
        inputTokens,
        outputTokens,
        totalTokens: inputTokens + outputTokens,
        estimatedCostUSD: 0,
      },
    };
  }
}
